zennit/rules.py

```diff
@@ -62,16 +62,31 @@ class AlphaBeta(LinearHook):
     beta: float, optional
         Multiplier for the negative output term.
     '''
-    def __init__(self, alpha=2., beta=1.):
+    def __init__(self, alpha=2., beta=1.): 
+        assert alpha > 0 and beta > 0
+        assert (alpha - beta) == 1
+
         super().__init__(
-            input_modifiers=[lambda input: input] * 2,
-            param_modifiers=[lambda param: param] * 2,
-            output_modifiers=[
-                lambda output: output.clamp(min=0),
-                lambda output: output.clamp(max=0),
-            ],
+            input_modifiers=[
+                lambda input: input.clamp(min=0), 
+                lambda input: input.clamp(max=0), 
+                lambda input: input.clamp(min=0), 
+                lambda input: input.clamp(max=0)  
+                ], 
+
+            param_modifiers=[
+                lambda param: param.clamp(min=0), 
+                lambda param: param.clamp(max=0), 
+                lambda param: param.clamp(max=0), 
+                lambda param: param.clamp(min=0) 
+                ],
+
+            output_modifiers=[lambda output: output] * 4,
+
             gradient_mapper=(lambda out_grad, outputs: [out_grad / stabilize(output) for output in outputs]),
-            reducer=(lambda inputs, gradients: inputs[0] * (alpha * gradients[0] + beta * gradients[1]))
+      
+            reducer=(lambda inputs, gradients: alpha * (inputs[0] * gradients[0] + inputs[1] * gradients[1]) 
+                - beta * (inputs[2] * gradients[2] + inputs[3] * gradients[3]))
```

Comment:
How is this different from using the output modifier instead?
The output-modifier uses the full gradient information, thus getting all these parts correctly.
Did you check its difference, and whether it produces the same output?

Reply:
The alpha beta rule is defined in http://iphome.hhi.de/samek/pdf/MonXAI19.pdf page 13

There you see, that we do not clamp the outputs but the inner part of the summation.
The matrix multiplication is defined as 
sum_i (input_i * weight_ij)
If you use output modifier you get
[sum_i (input_i * weight_ij)]⁺
but we want
sum_i [(input_i * weight_ij)]⁺

Hope it clarifies

Reply:
okay, agreed